method/stardust/stardust.r

```diff
@@ -181,93 +198,47 @@ if (technology %in% c("ST", "Visium")){
     spotPositions <- as.data.frame(SpatialExperiment::spatialCoords(spe)[,c(1,2)])
 }
 
-# Resolution optimization
-binary_search <- function(
-    countMatrix,
-    do_clustering,
-    n_clust_target,
-    resolution_update = 2,
-    resolution_init = 1,
-    resolution_boundaries=NULL,
-    num_rs = 100,
-    tolerance = 1e-3,
-    ...) {
-
-  # Initialize boundaries
-  lb <- rb <- NULL
-  n_clust <- -1
-
-  if (!is.null(resolution_boundaries)){
-    lb <- resolution_boundaries[1]
-    rb <- resolution_boundaries[2]
-  } else {
-    res <-  resolution_init
-    results <- do_clustering(countMatrix, res = res, ...)
-    # Adjust cluster_ids extraction per method
-    n_clust <- length(unique(results@active.ident))
-    if (n_clust > n_clust_target) {
-      while (n_clust > n_clust_target && res > 1e-5) {
-        rb <- res
-        res <- res/resolution_update
-        results <- do_clustering(countMatrix, res = res, ...)
-        n_clust <- length(unique(results@active.ident))
-      }
-      lb <- res
-    } else if (n_clust < n_clust_target) {
-      while (n_clust < n_clust_target) {
-        lb <- res 
-        res <- res*resolution_update
-        results <- do_clustering(countMatrix, res = res, ...)
-        n_clust <- length(unique(results@active.ident))
-      }
-      rb <- res
-    }
-    if (n_clust == n_clust_target) {lb = rb = res }
-  }
+# Clustering
+options(future.globals.maxSize = 100 * 1000 * 1024^2)
 
-  i <- 0
-  while ((rb - lb > tolerance || lb == rb) && i < num_rs) {
-    mid <- sqrt(lb * rb)
-    message("Resolution: ", mid)
-    set.seed(seed)
-    results <- do_clustering(countMatrix, res = mid, ...)
-    cluster_ids <- results@active.ident
-    # Adjust cluster_ids extraction per method
-    n_clust <- length(unique(cluster_ids))
-    message("Cluster: ", n_clust)
-    if (n_clust == n_clust_target || lb == rb) break
-    if (n_clust > n_clust_target) {
-      rb <- mid
-    } else {
-      lb <- mid
-    }
-    i <- i + 1
-  }
+countMatrix <- countMatrix[,sort(colnames(countMatrix))]
+d <- dim(spotPositions)[2]
+spotPositions <- spotPositions[,(d-1):d]
+spotPositions <- spotPositions[sort(rownames(spotPositions)),]
 
-  # Warning if target not met
-  if (n_clust != n_clust_target) {
-    warning(sprintf("Warning: n_clust = %d not found in binary search, return best approximation with res = %f and n_clust = %d. (rb = %f, lb = %f, i = %d)", n_clust_target, mid, n_clust, rb, lb, i))
-  }
-  return(results)
+seuratObject <- Seurat::CreateSeuratObject(countMatrix)
+seuratObject <- suppressWarnings(Seurat::SCTransform(seuratObject, assay = "RNA", verbose = FALSE))
```

Comment:
SCTransform has a parameter `variable.features.n` (default 3000) that resembles `n_genes`. This could be a place to adjust the number of variable features